### PR TITLE
Dialect/tsql update - added pivot / unpivot, view support, sequence support on table creation

### LINKED ISSUE #1467: TSQL PIVOT/UNPIVOT parsing error
PIVOT/UNPIVOT statements causes parsing error.
Syntax description: [from-using-pivot-and-unpivot](https://docs.microsoft.com/en-us/sql/t-sql/queries/from-using-pivot-and-unpivot?view=sql-server-ver15)

## Example
```sql
select 
	  tbl.column_name
	, [1] as Lvl_1 
	, [2] as Lvl_2
	, [3] as Lvl_3
from 
	table_name tbl
pivot
	(max(value) for rn in([1],[2],[3]) ) as pvt
```
parsing result:

```
[L:  7, P:  2]      |                from_expression:
[L:  7, P:  2]      |                    [META] indent:
[L:  7, P:  2]      |                    from_expression_element:
[L:  7, P:  2]      |                        table_expression:
[L:  7, P:  2]      |                            table_reference:
[L:  7, P:  2]      |                                identifier:                   'table_name'
[L:  7, P: 12]      |                        whitespace:                           ' '
[L:  7, P: 13]      |                        alias_expression:
[L:  7, P: 13]      |                            identifier:                       'tbl'
[L:  7, P: 16]      |                    newline:                                  '\n'
[L:  8, P:  1]      |                    [META] dedent:
[L:  8, P:  1]      |                unparsable:                                   !! Expected: 'Nothing...'
[L:  8, P:  1]      |                    raw:                                      'pivot'
[L:  8, P:  6]      |                    newline:                                  '\n'
```

## Expected Behaviour
No parsing errors, construction is parsed as special pivot segment

## Observed Behaviour
```
==== parsing violations ====
L:   8 | P:   1 |  PRS | Line 8, Position 1: Found unparsable section:
                       | 'pivot\n\t(max(value) for rn in([1],[2],[3]...'
```

## Dialect
`dialect = tsql`

## Version
`Python 3.9.6`
`sqlfluff, version 0.6.6`

### PR DESCRIPTION
**Pull request by ericmuijs**
### Brief summary of the change made
This PR extends SQL functionality such that I am able to parse nearly all views in a large SQL project. Including select with pivot and unpivot support.
Fixes #1467

### Are there any other side effects of this change that we should be aware of?
No

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

Added new SQL files with YAML to test functionality.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/parser` (note YML files can be auto generated with `python test/generate_parse_fixture_yml.py` or by running `tox` locally).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/dialects/tsql_keywords.py [hunk 1]
```diff
@@ -189,4 +189,44 @@
...
+    "SHOWPLAN_TEXT",
+    "SHOWPLAN_XML",
+    "STATISTICS IO",
+    "STATISTICS XML",
+    "STATISTICS PROFILE",
+    "STATISTICS TIME",
+    "IMPLICIT_TRANSACTIONS",
+    "REMOTE_PROC_TRANSACTIONS",
+    "TRANSACTION ISOLATION LEVEL",
+    "XACT_ABORT",
```
Comment (tunetheweb) [inline src/sqlfluff/dialects/tsql_keywords.py:231]: Would it make sense to sort these alphabetically? To make it easier to find if a word has been added or not?

File: test/fixtures/parser/tsql/create_table_with_sequence.sql [hunk 2]
```diff
@@ -0,0 +1,7 @@
+CREATE TABLE DEST.MELDER(
+      [ID] INT PRIMARY KEY NOT NULL DEFAULT NEXT VALUE FOR [dbo].[SEQ_MELDER]
+      ,[DOWNLOADED_TOTAL] INT NULL
+      ,[WARNED_DAILY] INT NULL
+      ,[DATE_OF_REPORT] DATETIME NULL
+      ,DATE_LAST_INSERTED DATETIME DEFAULT GETDATE()
+);
```
Comment (tunetheweb) [inline test/fixtures/parser/tsql/create_table_with_sequence.sql:7]: Nit can we add newlines? We have an `.editorconfig` file if you editor is compatible:

```suggestion
);

```

There's a few `.sql` files that have this issue.

File: test/fixtures/parser/tsql/stored_procedure_single_statement.sql [hunk 3]
```diff
@@ -35,35 +35,5 @@ BEGIN
        VWSSTAGE.ECDC_CASES
     WHERE DATE_LAST_INSERTED = (SELECT MAX(DATE_LAST_INSERTED) from VWSSTAGE.ECDC_CASES)
 
-    INSERT INTO VWSINTER.ECDC_CASES_OLD
-    (
-        [COUNTRY],
-        [COUNTRY_CODE],
```
Comment (tunetheweb) [inline test/fixtures/parser/tsql/stored_procedure_single_statement.sql:41]: Doesn't look like this really parsed entirely correctly, but still a shame to lose a test case that presumably was passing. Should this be copied into its own `insert_into_statement.sql` file?

**Review by tunetheweb — CHANGES_REQUESTED** [review]
Looks good. A few points of minor feedback!

File: src/sqlfluff/dialects/dialect_tsql.py [hunk 4]
```diff
@@ -77,6 +78,22 @@
...
+    PrimaryKeyGrammar=Sequence(
+        "PRIMARY", "KEY", OneOf("CLUSTERED", "NONCLUSTERED", optional=True)
+    ),
+    FromClauseTerminatorGrammar=OneOf(
+        "WHERE",
+        "LIMIT",
+        "GROUP",
+        "ORDER",
+        "HAVING",
+        "QUALIFY",
```
Comment (jpers36) [inline src/sqlfluff/dialects/dialect_tsql.py:90]: Do we need QUALIFY and WINDOW here?  Pretty sure they are not recognized in TSQL.

File: src/sqlfluff/dialects/dialect_tsql.py [hunk 5]
```diff
@@ -87,10 +104,183 @@ class StatementSegment(ansi_dialect.get_segment("StatementSegment")):  # type: i
...
+        Ref("SelectClauseSegment"),
+        # Dedent for the indent in the select clause.
+        # It's here so that it can come AFTER any whitespace.
+        Dedent,
+        Ref("FromClauseSegment", optional=True),
+        Ref("PivotUnpivotStatementSegment", optional=True),
+        Ref("WhereClauseSegment", optional=True),
+        Ref("GroupByClauseSegment", optional=True),
+        Ref("HavingClauseSegment", optional=True),
+        Ref("OverlapsClauseSegment", optional=True),
```
Comment (jpers36) [inline src/sqlfluff/dialects/dialect_tsql.py:149]: Do we need the OverlapsClauseSegment here?  Pretty sure it's not valid TSQL.

File: src/sqlfluff/dialects/dialect_tsql.py [hunk 6]
```diff
@@ -87,10 +104,183 @@ class StatementSegment(ansi_dialect.get_segment("StatementSegment")):  # type: i
...
+    https://docs.microsoft.com/en-us/sql/t-sql/statements/create-index-transact-sql?view=sql-server-ver15
+    """
+
+    type = "create_index_statement"
+    match_grammar = Sequence(
+        "CREATE",
+        Ref("OrReplaceGrammar", optional=True),
+        Sequence("UNIQUE", optional=True),
+        OneOf("CLUSTERED", "NONCLUSTERED", optional=True),
+        "INDEX",
```
Comment (jpers36) [inline src/sqlfluff/dialects/dialect_tsql.py:187]: Would it be feasible to add COLUMNSTORE to this, or should we hold that for a later improvement?

File: src/sqlfluff/dialects/dialect_tsql.py [hunk 7]
```diff
@@ -155,6 +345,79 @@ class DatatypeSegment(BaseSegment):
...
+
+    match_grammar = Sequence(
+        OneOf(
+            Sequence(Ref("IfNotExistsGrammar"), Ref("SelectStatementSegment")),
+            Sequence(Ref("IfExistsGrammar"), Ref("SelectStatementSegment")),
+            "IF",
+            Ref("ExpressionSegment"),
+        ),
+        Ref("StatementSegment"),
+        Sequence("ELSE", Ref("StatementSegment"), optional=True),
```
Comment (jpers36) [inline src/sqlfluff/dialects/dialect_tsql.py:378]: Can we get a test case for ELSE?  I couldn't get ELSE to parse correctly with a similar approach.

**Comment by tunetheweb** [comment]
Thanks for the review @jpers36 ! Great to have this oversight from someone who actually uses and understands T-SQL and hopefully @ericmuijs can address each of your queries.

File: src/sqlfluff/dialects/tsql_keywords.py (hunk 1 above)
Comment (ericmuijs) [reply src/sqlfluff/dialects/tsql_keywords.py]: Done

File: src/sqlfluff/dialects/dialect_tsql.py (hunk 6 above)
Comment (ericmuijs) [reply src/sqlfluff/dialects/dialect_tsql.py:187]: Later please, PR already big, feel free to add.

File: src/sqlfluff/dialects/dialect_tsql.py (hunk 4 above)
Comment (ericmuijs) [reply src/sqlfluff/dialects/dialect_tsql.py:90]: No we dont need them :)

File: src/sqlfluff/dialects/dialect_tsql.py (hunk 5 above)
Comment (ericmuijs) [reply src/sqlfluff/dialects/dialect_tsql.py:149]: Nope

File: src/sqlfluff/dialects/dialect_tsql.py (hunk 7 above)
Comment (ericmuijs) [reply src/sqlfluff/dialects/dialect_tsql.py:378]: If else with begin end etc is still pending. Was also not able to get it working yet.

**Review by tunetheweb — APPROVED** [review]

**Pull request merged**